components/style/sharing/mod.rs

```diff
@@ -175,10 +373,16 @@ impl<E: TElement> StyleSharingCandidateCache<E> {
 
         debug!("Inserting into cache: {:?} with parent {:?}", element, parent);
 
+        // If we know the element had no pres hints from the relations, we can
+        // just do as-if we had computed it, regardless of whether we actually
+        // did.
+        if !relations.intersects(AFFECTED_BY_PRESENTATIONAL_HINTS) {
```

Comment:
This seems like a hack. How about eliminating AFFECTED_BY_PRESENTATIONAL_HINTS altogether, and instead having push_applicable_declarations store the synthesis result directly in the StyleSharingCriteria hanging off the CurrentElementInfo?